README.md
What is this?
=============

Abacus is a tool for simple arithmetic calculations.

How to use?
===========

Abacus is not part of the Sugar desktop, but can be added.  Please refer to;

* [How to Get Sugar on sugarlabs.org](https://sugarlabs.org/),
* [How to use Sugar](https://help.sugarlabs.org/),
* [Download Abacus using Browse](https://activities.sugarlabs.org/), search for `Abacus`, then download, and;
* [How to use Abacus](https://help.sugarlabs.org/abacus.html)

How to upgrade?
===============

On Sugar desktop systems;
* use [My Settings](https://help.sugarlabs.org/my_settings.html), [Software Update](https://help.sugarlabs.org/my_settings.html#software-update), or;
* use Browse to open [activities.sugarlabs.org](https://activities.sugarlabs.org/), search for `Abacus`, then download.

How to integrate?
=================

On Fedora systems;

```
dnf install sugar-abacus
```

Abacus depends on Python, GTK+ 3, and the Sugar Toolkit.

Abacus is started by [Sugar](https://github.com/sugarlabs/sugar).


NEWS
62

* fix beads label to show string instead of bytes (Srevin Saju)
* Add numpad key events (Ibiam Chihurumnaya)

61

* New translations (Chris Leonard et al),
* Port to Python 3 (Rahul Bothra),
* Update README.md (James Cameron),
* Fixing files for deployment of ASLOv3 (Vipul Gupta),
* Add screenshots and update metadata (Ashnidh),

60.1

* Fix fail to start when LANG not valid (James Cameron),

60

* New translations (Chris Leonard et al),
* Remove suspicious code (Robert Antoni Buj Gelonch),
* Fix Python GI API version warnings (James Cameron),
* Update POT file (James Cameron),
* Remove duplicate LICENSE, use COPYING (James Cameron),
* Adjust README.md (James Cameron),
* Remove unused .abacusrc file  (James Cameron),

59

* New translations
* Added repository to activity.info

58

BUG FIX:
* Fixed clipboard for GTK3 (tch)

57
ENHANCEMENT:
* New translations (gn and ibo)

56

ENHANCEMENT:
* More speed ups (w/Gonzalo Odiard)
* New translations

BUG FIX:
* Properly restore custom abacus

55

ENHANCEMENT:
* Slight speedup on launch
* New translations

54

ENHANCEMENT:
* New translations

BUG FIX:
* Custom abacus autoselected when generated

53

ENHANCEMENT:
* Use busy cursor instead of alert when switching between abaci
* New translations

BUG FIX:
* Fixed bug preventing GNOME version from launching

51

ENHANCEMENT:
* Replace antiquated license strings

49

ENHANCEMENT:
* Increase label font size

47

BUG_FIX
* Add support for screen rotation

45

ENHANCEMENTS
* New translations

43

ENHANCEMENTS
* New translations
* Use lockable palette (Gonzalo Odiard)

41

ENHANCEMENT
* New translations

39

BUG_FIX
* Use set_icon_name as per #3976

37

ENHANCEMENT
* New translations
* Added summary to activity.info

35

ENHANCEMENT
* Using palette for spinners (#3563 and #3480)

BUG_FIX
* Remove work-around label color-style problem (#3479) in light of
  patch to fix #3558 and #3479

34

ENHANCEMENTS
* New translations
* Notify user of delay in loading new abacus
* Display abacus name on toolbar

BUG_FIXES
* Work-around label color-style problem (#3479)
* Refactoring to fix problem with excessive memory usage
* Don't allow custom abacus larger than screen
* Update GNOME version to use refactoring

33

ENHANCEMENT
* New translations

32

BUG FIX
* Update to latest sugar3-toolkit requirements (Gonzalo Oriard)

31

ENHANCEMENT
* GTK-3 conversion

30

UNDER THE HOOD
* Using improved Cairo graphics sprite library
* Using standard toolbar library

29

ENHANCEMENT
* Conversion to Cairo graphics

28

ENHANCEMENTS
* New translations
* Improved? labeling scheme

27

ENHANCEMENTS
* Use gear emblem instead of star emblem on custom abacus icon

BUG FIX
* Custom tool now updates radio button to indicate custom abacus is selected

26

ENHANCEMENTS
* Added button for selecting custom abacus
* Added colors to custom abacus

BUG FIX
* Fixed problem generating custom abacus under new refactoring

25

EHNANCEMENTS
* Using profile colors for bead colors
* Moved units of Soroban to center of abacus
* Added dots to Soroban
* Added reset button to toolbar and removed it from the abacus itself
* Move calculation to above the abacus rather than on the crossbar

24

ENHANCEMENTS
* New translations

23

ENHANCEMENTS
* New translations
* Updated icons

22

ENHANCEMENTS
* New translations for de, hy, ku, nah, nl, and tzm
* Updated toolbars and icons
* Label text uses ellipsis instead of shrinking

21

BUG FIX

* Fixed regression with old-style toolbars that preventing activity
  launch (#2942)

20

ENHANCEMENTS
* Icon cleanup
* White background

BUG FIXES
* Fixed problem with beads falling off abacus (#2933)
* Fixed problem with launching from GNOME shell

19

* Type in values, copy/paste from clipboard
* Abacus value propogates as you move between abaci

18

* Optimization of redraw code
* Spanish translations

17

* Fixed problem with moving multiple beads on decimal abacus
* Fixed hide problem with desktop version

16

* Added color to Decimal abacus

15

* Added "Caacupe" abacus
* Added Cuisenaire-like abacus 

14

* Simplifed toolbars

13

* Added save/restore for custom abacus

12

* Decimal abacus
* Reset button

11

* Performance improvement in label code

10

* Added labels to the beads (with inspiration and help from tuukkah)

9

* Customization toolbar -- design your own abacus

8

* Added hexadecimal
* Code cleanup to make it easier to modify/extend

7

* Fixed restore bug with fraction abacus
* Added missing type to rods

6

* Added binary abacus
* Added fraction abacus
* Autogenerate SVG
* General code clean-up/consolidation

5

* Highlight recently moved beads
* Fixed scaling problem with rods, dividing bar

4

* Save/restore abacus value
* Movable indicator

3

* Added frame
* Code clean up for easier maintenance

2

* Added Mayan (nepohualtzintzin) abacus
* Save/restore mode to/from Journal

1

* Abacus activity
* Chinese (suanpan), Japanese (soroban), and Russian (schety) models


abacus.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2010-14, Walter Bender

# This file is part of the Abacus Activity.

# The Abacus Activity is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# The Abacus Activity is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with the Abacus Activity.  If not, see <http://www.gnu.org/licenses/>.


import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, Gdk

from gettext import gettext as _

from abacus_window import Abacus


class AbacusMain:
    ABACI = {
        'b': 'binary',
        'c': 'suanpan',
        'f': 'fraction',
        'h': 'hexadecimal',
        'j': 'soroban',
        'm': 'nepohualtzintzin',
        'r': 'schety',
        'd': 'decimal',
        'C': 'caacupe',
        'R': 'cuisenaire'
    }

    def __init__(self):
        self.r = 0
        self.tw = None

        self.win = Gtk.Window(type=Gtk.WindowType.TOPLEVEL)
        self.x = 50
        self.y = 50
        self.width = 800
        self.height = 550
        self.win.set_default_size(self.width, self.height)
        self.win.move(self.x, self.y)
        self.win.maximize()
        self.win.set_title(_('Abacus'))
        self.win.connect('delete_event', lambda w, e: Gtk.main_quit())

        menu = Gtk.Menu()
        for k, v in self.ABACI.items():
            menu_items = Gtk.MenuItem.new_with_label(v)
            menu.append(menu_items)
            menu_items.connect('activate', self._switch_abacus_cb, k)
        menu_items = Gtk.MenuItem.new_with_label(_('Reset'))
        menu.append(menu_items)
        menu_items.connect('activate', self._reset)
        menu_items = Gtk.MenuItem.new_with_label(_('Quit'))
        menu.append(menu_items)
        menu_items.connect('activate', self.destroy)
        menu_items.show()
        root_menu = Gtk.MenuItem.new_with_label('Tools')
        root_menu.show()
        root_menu.set_submenu(menu)

        vbox = Gtk.VBox()
        self.win.add(vbox)
        vbox.show()

        menu_bar = Gtk.MenuBar()
        vbox.pack_start(menu_bar, False, False, 2)

        menu_bar.append(root_menu)
        menu_bar.show_all()

        canvas = Gtk.DrawingArea()
        width = Gdk.Screen.width()
        height = Gdk.Screen.height()
        canvas.set_size_request(width, height)
        vbox.pack_end(canvas, True, True, 0)

        self.win.show()

        self.abacus = Abacus(canvas)
        canvas.show()
        self.abacus.win = self.win
        self.abacus.activity = self
        self.abacus.init()

    def set_title(self, title):
        self.win.set_title(title)
        return

    def _switch_abacus_cb(self, widget, user):
        value = int(float(self.abacus.mode.value()))
        self.abacus.select_abacus(self.ABACI[user])
        self.abacus.mode.set_value_from_number(value)
        self.abacus.mode.label(self.abacus.generate_label())
        return True

    def _reset(self, event, data=None):
        ''' Reset beads to initial position '''
        self.abacus.mode.reset_abacus()
        self.abacus.mode.label(self.abacus.generate_label())
        return

    def destroy(self, event, data=None):
        ''' Callback for destroy event. '''
        Gtk.main_quit()


def main():
    Gtk.main()
    return 0


if __name__ == '__main__':
    AbacusMain()
    main()


po/pocreator.sh
#cp -r $1.pot  cpp.po  
#cp -r $1.pot  fil.po  
#cp -r $1.pot  ja.po  
#cp -r $1.pot  mvo.po  
#cp -r $1.pot  pt_BR.po  
cp -r $1.pot  sv.po   
cp -r $1.pot  ur.po
#cp -r $1.pot  af.po  
#cp -r $1.pot  cs.po   
#cp -r $1.pot  fr.po   
#cp -r $1.pot  km.po   
#cp -r $1.pot  na.po   
#cp -r $1.pot  qu.po     
#cp -r $1.pot  w.po   
cp -r $1.pot  vi.po
#cp -r $1.pot  am.po  
cp -r $1.pot  de.po   
cp -r $1.pot  gu.po   
#cp -r $1.pot  kn.po   
cp -r $1.pot  nb.po   
#cp -r $1.pot  ro.po     
#cp -r $1.pot  ta.po   
#cp -r $1.pot  wa.po
#cp -r $1.pot  ar.po  
#cp -r $1.pot  en.po
#cp -r $1.pot  hi.po
#cp -r $1.pot  lv.po
#cp -r $1.pot  zh_CN.po
cp -r $1.pot  th.po 
cp -r $1.pot  rw.po
cp -r $1.pot  nl.po
cp -r $1.pot  kos.po
cp -r $1.pot  he.po
cp -r $1.pot  el.po
#cp -r $1.pot  yo.po
#cp -r $1.pot  te.po
cp -r $1.pot  ru.po
#cp -r $1.pot  ne.po
cp -r $1.pot  ko.po
#cp -r $1.pot  ha.po
#cp -r $1.pot  dz.po    
#cp -r $1.pot  ar_SY.po  
cp -r $1.pot  ay.po       
cp -r $1.pot  sd.po
cp -r $1.pot  ton.po
#cp -r $1.pot  zh_TW.po
cp -r $1.pot  pis.po
cp -r $1.pot  sl.po
#cp -r $1.pot  tvl.po
#cp -r $1.pot  fa_AF.po
#cp -r $1.pot ig.po 
#cp -r $1.pot nn.po
cp -r $1.pot pis.po
#cp -r $1.pot ml.po 
#cp -r $1.pot  hy.po 
#cp -r $1.pot fa.po 
cp -r $1.pot tr.po
#cp -r $1.pot sk.po
#cp -r $1.pot pap.po
#cp -r $1.pot mk.po
cp -r $1.pot hu.po
#cp -r $1.pot es.po
#cp -r $1.pot tpi.po
cp -r $1.pot si.po
cp -r $1.pot pa.po
cp -r $1.pot mg.po
#cp -r $1.pot ht.po
#cp -r $1.pot en_US.po
#cp -r $1.pot bg.po    
#cp -r $1.pot bi.po   
#cp -r $1.pot bn.po   
cp -r $1.pot bn_IN.po     
#cp -r $1.pot bs.po    
#cp -r $1.pot ca.po   
#cp -r $1.pot uk.po
#cp -r $1.pot st.po
#cp -r $1.pot pt.po
#cp -r $1.pot ms.po
#cp -r $1.pot it.po
#cp -r $1.pot fi.po
cp -r $1.pot ug.po
cp -r $1.pot sq.po
cp -r $1.pot ps.po
#cp -r $1.pot  mr.po
cp -r $1.pot is.po
cp -r $1.pot ff.po
#cp -r $1.pot tzo.po
#cp -r $1.pot sm.po
cp -r $1.pot pl.po
#cp -r $1.pot mn.po 






toolbar_utils.py
# -*- coding: utf-8 -*-
# Copyright (c) 2011, Walter Bender

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# You should have received a copy of the GNU General Public License
# along with this library; if not, write to the Free Software
# Foundation, 51 Franklin Street, Suite 500 Boston, MA 02110-1335 USA


from gi.repository import Gtk

from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics.radiotoolbutton import RadioToolButton


def button_factory(icon_name, toolbar, callback, cb_arg=None, tooltip=None,
                   accelerator=None):
    ''' Factory for making toolbar buttons '''
    button = ToolButton(icon_name)
    if tooltip is not None:
        button.set_tooltip(tooltip)
    button.props.sensitive = True
    if accelerator is not None:
        button.props.accelerator = accelerator
    if cb_arg is None:
        button.connect('clicked', callback)
    else:
        button.connect('clicked', cb_arg)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(button, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(button, -1)
    button.show()
    return button


def radio_factory(icon_name, toolbar, callback, cb_arg=None,
                  tooltip=None, group=None):
    ''' Add a radio button to a toolbar '''
    button = RadioToolButton(group=group)
    button.set_icon_name(icon_name)
    if tooltip is not None:
        button.set_tooltip(tooltip)
    if cb_arg is None:
        button.connect('clicked', callback)
    else:
        button.connect('clicked', callback, cb_arg)
    if hasattr(toolbar, 'insert'):  # the main toolbar
        toolbar.insert(button, -1)
    else:  # or a secondary toolbar
        toolbar.props.page.insert(button, -1)
    button.show()
    return button


def label_factory(label_text, toolbar):
    ''' Factory for adding a label to a toolbar '''
    label = Gtk.Label(label=label_text)
    label.set_line_wrap(True)
    label.show()
    toolitem = Gtk.ToolItem()
    toolitem.add(label)
    toolbar.insert(toolitem, -1)
    toolitem.show()
    return label


def spin_factory(default, min_value, max_value, callback, toolbar):
    ''' Factory for making toolbar value spinners '''
    spin_adj = Gtk.Adjustment(default, min_value, max_value, 1, 32, 0)
    spin = Gtk.SpinButton()
    spin.set_adjustment(spin_adj)
    spin.connect('value-changed', callback)
    spin.set_numeric(True)
    spin.show()
    toolitem = Gtk.ToolItem()
    toolitem.add(spin)
    if toolbar is not None:
        toolbar.insert(toolitem, -1)
        toolitem.show()
        return spin
    else:
        toolitem.show()
        return spin, toolitem


def separator_factory(toolbar, expand=False, visible=True):
    ''' Add a separator to a toolbar '''
    separator = Gtk.SeparatorToolItem()
    separator.props.draw = visible
    separator.set_expand(expand)
    toolbar.insert(separator, -1)
    separator.show()
    return separator


AbacusActivity.py
# -*- coding: utf-8 -*-
# Copyright 2010-13, Walter Bender

# This file is part of the Abacus Activity.

# The Abacus Activity is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# The Abacus Activity is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with the Abacus Activity.  If not, see <http://www.gnu.org/licenses/>.


import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import GObject

from sugar3.activity import activity
from sugar3 import profile
from sugar3.graphics.toolbarbox import ToolbarBox
from sugar3.activity.widgets import ActivityToolbarButton
from sugar3.activity.widgets import StopButton
from sugar3.graphics.toolbarbox import ToolbarButton
from sugar3.graphics.toolbutton import ToolButton
from sugar3.graphics import style

from gettext import gettext as _

import logging
_logger = logging.getLogger('abacus-activity')

from abacus_window import Abacus, Custom, MAX_RODS, MAX_TOP, MAX_BOT
from toolbar_utils import separator_factory, radio_factory, label_factory, \
    button_factory, spin_factory


NAMES = {
    # TRANS: http://en.wikipedia.org/wiki/Suanpan (Chinese abacus)
    'suanpan': _('Suanpan'),
    # TRANS: http://en.wikipedia.org/wiki/Soroban (Japanese abacus)
    'soroban': _('Soroban'),
    # TRANS: simple decimal abacus
    'decimal': _('Decimal'),
    # TRANS: http://en.wikipedia.org/wiki/Abacus#Native_American_abaci
    'nepohualtzintzin': _('Nepohualtzintzin'),
    # TRANS: hexidecimal abacus
    'hexadecimal': _('Hexadecimal'),
    # TRANS: binary abacus
    'binary': _('Binary'),
    # TRANS: http://en.wikipedia.org/wiki/Abacus#Russian_abacus
    'schety': _('Schety'),
    # TRANS: abacus for adding fractions
    # 'fraction': _('Fraction'),
    # TRANS: Abacus invented by teachers in Caacupé, Paraguay
    'caacupe': _('Caacupé'),
    # TRANS: Cuisenaire Rods
    'cuisenaire': _('Rods'),
    # TRANS: Custom abacus
    'custom': _('Custom')
}


class AbacusActivity(activity.Activity):

    def __init__(self, handle):
        ''' Initiate activity. '''
        super(AbacusActivity, self).__init__(handle)

        self._setting_up = True
        self.bead_colors = profile.get_color().to_string().split(',')

        # no sharing
        self.max_participants = 1

        self.sep = []
        self.abacus_toolbar = Gtk.Toolbar()
        custom_toolbar = Gtk.Toolbar()
        edit_toolbar = Gtk.Toolbar()

        toolbox = ToolbarBox()

        activity_button = ActivityToolbarButton(self)
        toolbox.toolbar.insert(activity_button, 0)
        activity_button.show()

        edit_toolbar_button = ToolbarButton(label=_('Edit'),
                                            page=edit_toolbar,
                                            icon_name='toolbar-edit')
        edit_toolbar_button.show()
        toolbox.toolbar.insert(edit_toolbar_button, -1)
        edit_toolbar_button.show()

        self.abacus_toolbar_button = ToolbarButton(
            page=self.abacus_toolbar,
            icon_name='abacus-list')
        self.abacus_toolbar.show()
        toolbox.toolbar.insert(self.abacus_toolbar_button, -1)
        self.abacus_toolbar_button.show()

        self.custom_toolbar_button = ToolbarButton(
            page=custom_toolbar,
            icon_name='view-source')
        custom_toolbar.show()
        toolbox.toolbar.insert(self.custom_toolbar_button, -1)
        self.custom_toolbar_button.show()

        separator_factory(toolbox.toolbar, False, True)

        button_factory('edit-delete', toolbox.toolbar,
                       self._reset_cb, tooltip=_('Reset'))

        separator_factory(toolbox.toolbar, False, True)

        self._label = label_factory(NAMES['suanpan'], toolbox.toolbar)

        separator_factory(toolbox.toolbar, True, False)

        stop_button = StopButton(self)
        stop_button.props.accelerator = _('<Ctrl>Q')
        toolbox.toolbar.insert(stop_button, -1)
        stop_button.show()

        self.set_toolbar_box(toolbox)
        toolbox.show()

        self.abacus_buttons = {}

        # Traditional
        self._add_abacus_button('decimal', None)
        self._add_abacus_button('soroban', self.abacus_buttons['decimal'])
        self._add_abacus_button('suanpan', self.abacus_buttons['decimal'])

        self.sep.append(separator_factory(self.abacus_toolbar))

        # Bases other than 10
        self._add_abacus_button('nepohualtzintzin',
                                self.abacus_buttons['decimal'])
        self._add_abacus_button('hexadecimal', self.abacus_buttons['decimal'])
        self._add_abacus_button('binary', self.abacus_buttons['decimal'])

        self.sep.append(separator_factory(self.abacus_toolbar))

        # Fractions
        self._add_abacus_button('schety', self.abacus_buttons['decimal'])
        # self._add_abacus_button('fraction', self.abacus_buttons['decimal'])
        self._add_abacus_button('caacupe', self.abacus_buttons['decimal'])

        self.sep.append(separator_factory(self.abacus_toolbar))

        # Non-traditional
        self._add_abacus_button('cuisenaire', self.abacus_buttons['decimal'])

        self.sep.append(separator_factory(self.abacus_toolbar))

        # Custom
        self._add_abacus_button('custom', self.abacus_buttons['decimal'])

        preferences_button = ToolButton('preferences-system')
        preferences_button.set_tooltip(_('Custom'))
        custom_toolbar.insert(preferences_button, -1)
        preferences_button.palette_invoker.props.toggle_palette = True
        preferences_button.palette_invoker.props.lock_palette = True
        preferences_button.props.hide_tooltip_on_click = False
        preferences_button.show()

        self._palette = preferences_button.get_palette()
        button_box = Gtk.VBox()
        # TRANS: Number of rods on the abacus
        self._rods_spin = add_spinner_and_label(
            15, 1, MAX_RODS, _('Rods:'), self._rods_spin_cb, button_box)
        # TRANS: Number of beads in the top section of the abacus
        self._top_spin = add_spinner_and_label(
            2, 0, MAX_TOP, _('Top:'), self._top_spin_cb, button_box)
        # TRANS: Number of beads in the bottom section of the abacus
        self._bottom_spin = add_spinner_and_label(
            5, 0, MAX_BOT, _('Bottom:'), self._bottom_spin_cb, button_box)
        # TRANS: Scale factor between bottom and top beads
        self._value_spin = add_spinner_and_label(
            5, 1, MAX_BOT + 1, _('Factor:'), self._value_spin_cb, button_box)
        # TRANS: Scale factor between rods
        self._base_spin = add_spinner_and_label(
            10, 1, (MAX_TOP + 1) * MAX_BOT, _('Base:'), self._base_spin_cb,
            button_box)
        hbox = Gtk.HBox()
        hbox.pack_start(button_box, True, True, style.DEFAULT_SPACING)
        hbox.show_all()
        self._palette.set_content(hbox)

        separator_factory(custom_toolbar, False, False)

        self.custom_maker = button_factory('new-abacus', custom_toolbar,
                                           self._custom_cb,
                                           tooltip=_('Custom'))

        button_factory('edit-copy', edit_toolbar, self._copy_cb,
                       tooltip=_('Copy'), accelerator='<Ctrl>c')
        button_factory('edit-paste', edit_toolbar, self._paste_cb,
                       tooltip=_('Paste'), accelerator='<Ctrl>v')

        # Create a canvas
        canvas = Gtk.DrawingArea()
        canvas.set_size_request(Gdk.Screen.width(),
                                Gdk.Screen.height())
        self.set_canvas(canvas)
        canvas.show()
        self.show_all()

        # Initialize the canvas
        self.abacus = Abacus(canvas, self)

        self._setting_up = False

        # Read the current mode from the Journal
        if 'rods' in self.metadata:
            self._rods_spin.set_value(int(self.metadata['rods']))
        if 'top' in self.metadata:
            self._top_spin.set_value(int(self.metadata['top']))
        if 'bottom' in self.metadata:
            self._bottom_spin.set_value(int(self.metadata['bottom']))
        if 'factor' in self.metadata:
            self._value_spin.set_value(int(self.metadata['factor']))
        if 'base' in self.metadata:
            self._base_spin.set_value(int(self.metadata['base']))
        if 'abacus' in self.metadata:
            if self.metadata['abacus'] in self.abacus_buttons:
                _logger.debug('restoring %s', self.metadata['abacus'])
                if self.metadata['abacus'] == 'custom':
                    self._custom_cb()
                self.abacus_buttons[self.metadata['abacus']].set_active(True)
            else:  # Default is Chinese
                self.abacus_buttons['suanpan'].set_active(True)

            if 'value' in self.metadata:
                _logger.debug('restoring value %s', self.metadata['value'])
                self.abacus.mode.set_value(self.metadata['value'])
                self.abacus.mode.label(self.abacus.generate_label())

        self.abacus.init()

        # Start with abacus toolbar expanded and suanpan as default
        self.abacus_toolbar_button.set_expanded(True)

    def _add_abacus_button(self, name, group):
        self.abacus_buttons[name] = radio_factory(
            name,
            self.abacus_toolbar,
            self._radio_cb,
            cb_arg=name,
            tooltip=NAMES[name],
            group=group)

    def _radio_cb(self, button, abacus):
        self._select_abacus(abacus)

    def _reset_cb(self, button=None):
        self.abacus.mode.reset_abacus()
        self.abacus.mode.label(self.abacus.generate_label())

    def _notify_new_abacus(self, prompt):
        ''' Loading a new abacus can be slooow, so alert the user. '''
        # a busy cursor is adequate
        self.get_window().set_cursor(Gdk.Cursor.new(Gdk.CursorType.WATCH))

    def _select_abacus(self, abacus):
        ''' Notify the user of an expected delay and then... '''
        if not hasattr(self, 'abacus') or self._setting_up:
            _logger.debug('setting up')
            return
        # Not selected?
        if not self.abacus_buttons[abacus].get_active():
            _logger.debug('%s not active' % abacus)
            return

        self._notify_new_abacus(NAMES[abacus])
        # Give the cursor/alert time to load
        GObject.idle_add(self._switch_modes, abacus)

    def _switch_modes(self, abacus):
        ''' Display the selected abacus '''
        _logger.debug('switching modes to %s', abacus)
        if abacus == self.abacus.mode.name:
            _logger.debug('do not switch already in the same mode')
            self.get_window().set_cursor(None)
            return
        # Save current value
        value = int(float(self.abacus.mode.value()))
        if abacus == 'custom' and self.abacus.custom is None:
            self.custom_toolbar_button.set_expanded(True)
            # self.abacus.mode = self.abacus.custom
            self.get_window().set_cursor(None)
        else:
            _logger.debug('switch_mode: setting abacus to %s' % abacus)
            self.abacus.select_abacus(abacus)
            # Load saved value
            self.abacus.mode.set_value_from_number(value)
            self.abacus.mode.label(self.abacus.generate_label())
            self._label.set_text(NAMES[abacus])
            self.get_window().set_cursor(None)

    def _rods_spin_cb(self, button=None):
        return

    def _top_spin_cb(self, button=None):
        return

    def _bottom_spin_cb(self, button=None):
        return

    def _value_spin_cb(self, button=None):
        return

    def _base_spin_cb(self, button=None):
        return

    def _custom_cb(self, button=None):
        ''' Display the custom abacus; hide the others '''
        value = float(self.abacus.mode.value(count_beads=False))
        self.abacus.mode.hide()
        if self.abacus.custom is not None:
            self.abacus.custom.hide()
        self.abacus.custom = Custom(self.abacus, self.abacus.bead_colors)
        self.abacus.custom.set_custom_parameters(
            rods=self._rods_spin.get_value_as_int(),
            top=self._top_spin.get_value_as_int(),
            bot=self._bottom_spin.get_value_as_int(),
            factor=self._value_spin.get_value_as_int(),
            base=self._base_spin.get_value_as_int())
        self.abacus.custom.create()
        self.abacus.custom.draw_rods_and_beads()
        self.abacus.custom.show()
        self._label.set_text(NAMES['custom'])
        self.abacus.mode = self.abacus.custom
        self.abacus.mode_dict['custom'][0] = self.abacus.custom
        self.abacus_toolbar_button.set_expanded(True)
        self.abacus.mode.set_value_from_number(value)

    def _copy_cb(self, arg=None):
        ''' Copy a number to the clipboard from the active abacus. '''
        clipboard = Gtk.Clipboard.get(Gdk.SELECTION_CLIPBOARD)
        text = self.abacus.generate_label(sum_only=True)
        if text is not None:
            clipboard.set_text(text, -1)
        return

    def _paste_cb(self, arg=None):
        ''' Paste a number from the clipboard to the active abacus. '''
        clipboard = Gtk.Clipboard.get(Gdk.SELECTION_CLIPBOARD)
        text = clipboard.wait_for_text()
        if text is not None:
            try:
                self.abacus.mode.set_value_from_number(float(text))
            except ValueError as e:
                _logger.debug(str(e))
                return
            self.abacus.mode.label(self.abacus.generate_label())
        return

    def write_file(self, file_path):
        ''' Write the bead positions to the Journal '''
        _logger.debug('Saving current abacus to Journal: %s %s' % (
            self.abacus.mode.name, self.abacus.mode.value(True)))
        self.metadata['abacus'] = self.abacus.mode.name
        self.metadata['value'] = self.abacus.mode.value(True)
        self.metadata['rods'] = str(self._rods_spin.get_value_as_int())
        self.metadata['top'] = str(self._top_spin.get_value_as_int())
        self.metadata['bottom'] = str(self._bottom_spin.get_value_as_int())
        self.metadata['factor'] = str(self._value_spin.get_value_as_int())
        self.metadata['base'] = str(self._base_spin.get_value_as_int())


def add_spinner_and_label(default_value, min_value, max_value,
                          tooltip, cb, box):
    ''' Add a spinner and a label to a box '''
    spinner_and_label = Gtk.HBox()
    spinner, item = spin_factory(default_value, min_value, max_value, cb, None)
    label = Gtk.Label(label=tooltip)
    label.set_justify(Gtk.Justification.LEFT)
    label.set_line_wrap(True)
    label.show()
    spinner_and_label.pack_start(label, expand=False, fill=False, padding=0)
    label = Gtk.Label(label=' ')
    label.show()
    spinner_and_label.pack_start(label, expand=True, fill=False, padding=0)
    spinner_and_label.pack_start(item, expand=False, fill=False, padding=0)
    box.pack_start(spinner_and_label, expand=False, fill=False, padding=5)
    spinner_and_label.show()
    return spinner
